tests/core/arch-arm64/arm64stalker.c

```diff
@@ -0,0 +1,652 @@
+{
+  invoke_flat (fixture, GUM_NOTHING);
+  g_assert_cmpuint (fixture->sink->events->len, ==, 0);
+}
+
+STALKER_ARM64_TESTCASE (call)
+{
+  StalkerTestFunc func;
+  GumCallEvent *ev;
+
+  func = invoke_flat (fixture, GUM_CALL);
+
+  g_assert_cmpuint (fixture->sink->events->len, ==, 2);
+  g_assert_cmpint (g_array_index (fixture->sink->events, GumEvent,
+      0).type, ==, GUM_CALL);
+  ev = &g_array_index (fixture->sink->events, GumEvent, 0).call;
+  GUM_ASSERT_CMPADDR (ev->location, ==, fixture->last_invoke_calladdr);
+  GUM_ASSERT_CMPADDR (ev->target, ==, func);
+}
+
+STALKER_ARM64_TESTCASE (ret)
+{
+  StalkerTestFunc func;
+  GumRetEvent *ev;
+
+  func = invoke_flat (fixture, GUM_RET);
+
+  g_assert_cmpuint (fixture->sink->events->len, ==, 1);
+  g_assert_cmpint (g_array_index (fixture->sink->events, GumEvent,
+      0).type, ==, GUM_RET);
+
+  ev = &g_array_index (fixture->sink->events, GumEvent, 0).ret;
+
+  GUM_ASSERT_CMPADDR (ev->location, ==, ((guint8 *) GSIZE_TO_POINTER (
+      func)) + 3 * 4);
+  GUM_ASSERT_CMPADDR (ev->target, ==, fixture->last_invoke_retaddr);
+}
+
+STALKER_ARM64_TESTCASE (exec)
+{
+  StalkerTestFunc func;
+  GumRetEvent *ev;
+
+  func = invoke_flat (fixture, GUM_EXEC);
+
+  g_assert_cmpuint (fixture->sink->events->len, ==, INVOKER_INSN_COUNT + 4);
+  g_assert_cmpint (g_array_index (fixture->sink->events, GumEvent,
+      INVOKER_IMPL_OFFSET).type, ==, GUM_EXEC);
+  ev =
+      &g_array_index (fixture->sink->events, GumEvent, INVOKER_IMPL_OFFSET).ret;
+  GUM_ASSERT_CMPADDR (ev->location, ==, func);
+}
+
+STALKER_ARM64_TESTCASE (unconditional_branch)
+{
+  guint8 *code;
+  GumArm64Writer cw;
+  gpointer address;
+  const gchar *my_ken_lbl = "my_ken";
+  StalkerTestFunc func;
+
+  code = gum_alloc_n_pages (1, GUM_PAGE_RWX);
+  gum_arm64_writer_init (&cw, code);
+
+  gum_arm64_writer_put_push_all_registers (&cw);
+  gum_arm64_writer_put_call_address_with_arguments (&cw,
+      gum_stalker_follow_me, 2,
+      GUM_ARG_ADDRESS, fixture->stalker,
+      GUM_ARG_ADDRESS, fixture->sink);
+
+  gum_arm64_writer_put_pop_all_registers (&cw);
+
+  gum_arm64_writer_put_nop (&cw);
+  gum_arm64_writer_put_nop (&cw);
+  gum_arm64_writer_put_b_label (&cw, my_ken_lbl);
+
+  address = gum_arm64_writer_cur (&cw);
+  gum_arm64_writer_put_add_reg_reg_imm (&cw, ARM64_REG_X0, ARM64_REG_X0, 10);
+
+  gum_arm64_writer_put_push_all_registers (&cw);
+  gum_arm64_writer_put_call_address_with_arguments (&cw,
+      gum_stalker_unfollow_me, 1,
+      GUM_ARG_ADDRESS, fixture->stalker);
+  gum_arm64_writer_put_pop_all_registers (&cw);
+
+  gum_arm64_writer_put_ret (&cw);
+
+  gum_arm64_writer_put_label (&cw, my_ken_lbl);
+  gum_arm64_writer_put_add_reg_reg_imm (&cw, ARM64_REG_X0, ARM64_REG_X0, 1);
+  gum_arm64_writer_put_b_imm (&cw, address);
+
+  gum_arm64_writer_free (&cw);
+
+  fixture->sink->mask = GUM_CALL | GUM_RET | GUM_EXEC;
+  func = GUM_POINTER_TO_FUNCPTR (StalkerTestFunc, code);
+  int r = func (2);
+
+  g_assert (13 == r);
```

Comment:
Constant should always be on the right hand side (not Yoda-style like this :) ). Also, `g_assert_cmpint` would be better.